### DISCUSSION TITLE
Ratatui Tutor Binary

### DISCUSSION THREAD

**Discussion by orhun** [ratatui/ratatui#918] [Ideas]
The idea is:

```sh
$ cargo install ratatui
```

This will install a binary that displays a Ratatui tutorial in the terminal when you run it. The binary can be called `ratatui` or `ratatui-tutorial` (or `ratatutorial` lol) and it will be also built with Ratatui.

i.e. `vimtutor` but for Ratatui

For that, we can simply render our docs (markdown files) and make them _browsable_ in the terminal.

When it comes to the `cargo install` command we just need to add `main.rs` file to the repo for cargo to locate the binary to install. I'm not sure if there are any downsides to having both lib and binary in the same crate.

Thanks @mre for the idea!

**Comment by joshka** [comment]
I can't recall the exact downside (only that there was one annoying one). I've often seen people add a secondary crate - e.g. XXX-cli for this sort of thing.

Off the top of my head, there's no easy way to make the libs for the binary only affect the binary and not the crate. That's possibly not the exact problem, but it's something like that.

**Reply by joshka** [reply]
> Off the top of my head, there's no easy way to make the libs for the binary only affect the binary and not the crate. That's possibly not the exact problem, but it's something like that.

That is the main problem. If you have a library and a binary cli app in the same crate, then there's no good way to be able to make the packages that are required to run the binary not be required by all the users of the library. E.g. you might for instance use `anyhow` for error handling in the binary, but this shouldn't end up in the library.

That is to say. Make this `Ratatui-tutor` instead or similar.

**Reply by orhun** [reply]
I will take a stab at this soon. It was in my mind for a while - let's make it happen one way or another :)

Edit: Time constraints... feel free to experiment with this if you'd like.
